jerry-core/ecma/builtin-objects/ecma-builtin-reflect.c

```diff
@@ -0,0 +1,187 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <math.h>
+#include "ecma-builtins.h"
+#include "ecma-gc.h"
+#include "ecma-exceptions.h"
+#include "ecma-builtin-object.h"
+#include "ecma-function-object.h"
+#include "ecma-builtin-function-prototype.h"
+#include "jcontext.h"
+
+#if ENABLED (JERRY_ES2015_BUILTIN_REFLECT)
+
+#define ECMA_BUILTINS_INTERNAL
+#include "ecma-builtins-internal.h"
+
+/**
+ * This object has a custom dispatch function.
+ */
+#define BUILTIN_CUSTOM_DISPATCH
+
+/**
+ * List of built-in routine identifiers.
+ */
+enum
+{
+  ECMA_REFLECT_OBJECT_ROUTINE_START = ECMA_BUILTIN_ID__COUNT - 1,
+  ECMA_REFLECT_OBJECT_APPLY, /* ECMA-262 v6, 26.1.1 */
+  ECMA_REFLECT_OBJECT_DEFINE_PROPERTY, /* ECMA-262 v6, 26.1.3 */
+  ECMA_REFLECT_OBJECT_GET_OWN_PROPERTY_DESCRIPTOR_UL, /* ECMA-262 v5, 26.1.7 */
+  ECMA_REFLECT_OBJECT_GET_PROTOTYPE_OF_UL, /* ECMA-262 v6, 26.1.8 */
+  ECMA_REFLECT_OBJECT_IS_EXTENSIBLE, /* ECMA-262 v6, 26.1.10 */
+  ECMA_REFLECT_OBJECT_PREVENT_EXTENSIONS_UL, /* ECMA-262 v6, 26.1.12 */
+  ECMA_REFLECT_OBJECT_SET_PROTOTYPE_OF_UL, /* ECMA-262 v6, 26.1.14 */
+};
+
+#define BUILTIN_INC_HEADER_NAME "ecma-builtin-reflect.inc.h"
+#define BUILTIN_UNDERSCORED_ID reflect
+#include "ecma-builtin-internal-routines-template.inc.h"
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmabuiltins
+ * @{
+ *
+ * \addtogroup object ECMA Reflect object built-in
+ * @{
+ */
+
+/**
+ * Dispatcher for the built-in's routines.
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+ecma_value_t
+ecma_builtin_reflect_dispatch_routine (uint16_t builtin_routine_id,
+                                       ecma_value_t this_arg,
+                                       const ecma_value_t arguments_list[],
+                                       ecma_length_t arguments_number)
+{
+  JERRY_UNUSED (this_arg);
+  JERRY_UNUSED (arguments_number);
+
+  switch (builtin_routine_id)
+  {
+    case ECMA_REFLECT_OBJECT_APPLY:
+    {
+      if (!ecma_op_is_callable (arguments_list[0]))
+      {
+        return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a function."));
+      }
+      if (!ecma_is_value_object (arguments_list[0]))
+      {
+        return ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not Object."));
+      }
+      ecma_object_t *func_obj_p = ecma_get_object_from_value (arguments_list[0]);
+      return ecma_builtin_function_prototype_object_apply (func_obj_p, arguments_list[1], arguments_list[2]);
+    }
+    case ECMA_REFLECT_OBJECT_DEFINE_PROPERTY:
+    {
+      if (!ecma_is_value_object (arguments_list[0]))
```

Comment:
In almost all of the cases arguments_list[0] must be an object.
Can you make an early if check with the exception of get- and set prototype of functions, so the same code doesn't get repeated numerous times?